### ISSUE TITLE
2020/2/29

### ISSUE THREAD

**Issue by mingxoxo** [mingxoxo/deapLearning-study#4]
**view 함수(중요!)**
shape를 자유자재로 바꿀 수 있어야 한다.
![image](https://user-images.githubusercontent.com/44529556/75600162-78e24600-5aef-11ea-9943-faf9aad4db7a.png)

```
import numpy as np
import torch

t = np.array([[[0,1,2],
               [3,4,5]],
              [[6,7,8],
               [9,10,11]]])
ft = torch.FloatTensor(t)
print(ft.shape)

print(ft.view([-1,3]))
print(ft.view([-1,3]).shape)

print(ft.view([-1,1, 3]))
print(ft.view([-1,1,3]).shape)
```

**squeeze 함수**
dimension의 element의 갯수가 1일때 없애줌
![image](https://user-images.githubusercontent.com/44529556/75600293-e478e300-5af0-11ea-8aef-65f91e1f1f36.png)
squeeze(dim=?) -> 이 dim이 1이면 없애줌

**unsqueeze 함수**
 원하는 dim에 1을 넣어줌

**Type Casting 함수**
![image](https://user-images.githubusercontent.com/44529556/75600368-f14a0680-5af1-11ea-972d-d0d977babe33.png)

**Concatenate 함수**
이어붙이기! ->c언어에도 있는 그거
![image](https://user-images.githubusercontent.com/44529556/75600416-72090280-5af2-11ea-942a-f45f1d2db08e.png)

**Stacking 함수**
Concatenate을 편리하게 이용하게 단축시켜 놓은 것
![image](https://user-images.githubusercontent.com/44529556/75600461-f065a480-5af2-11ea-8bd6-3b4f2c0ccfe3.png)

**Ones and Zeros**
똑같은 shape안에 0과 1을 넣는?
같은 디바이스안에 텐서를 선언해준다
![image](https://user-images.githubusercontent.com/44529556/75600516-8a2d5180-5af3-11ea-8adf-b9b13b70cf1e.png)
```
x = torch.FloatTensor([[0,1,2],[2,1,0]])
print(torch.ones_like(x))
print(torch.zeros_like(x))
```
**In-place Operation**
새 tensor에 선언하지말고 기존의 tensor의 값을 변경하는 역할
![image](https://user-images.githubusercontent.com/44529556/75600562-fe67f500-5af3-11ea-8db4-9771e10ae944.png)

**Comment by mingxoxo** [comment]
학습시키는 데이터 training dataset / 검사 test dataset
![image](https://user-images.githubusercontent.com/44529556/75600597-7504f280-5af4-11ea-8066-709800a1a765.png)
입력데이터 x 출력데이터 y 각각의 tensor로 만든다.
![image](https://user-images.githubusercontent.com/44529556/75600620-c2815f80-5af4-11ea-8dd8-c7e419bd10aa.png)

Linear regression은 학습데이터와 가장 잘 맞는 하나의 직선을 찾는 일이다. 이 직선 y는 y = Wx + b의 형태로 나타낼 수 있다.
W는 weight , b는 Bias라고 한다.
처음에는 0으로 초기화시키고 W와 b를 찾는것이 우리의 목적이므로 requires_grad = True 를 통해서 학습할 것이라고 파이토치에게 알려준다.
![image](https://user-images.githubusercontent.com/44529556/75600690-a8944c80-5af5-11ea-8e80-12218a62fb32.png)

**Compute loss**
얼마나 정답과 가까운지 알아야 함-> cost 또는 loss
loss -> Mean Squared Error(MSE)라고 부른다
![image](https://user-images.githubusercontent.com/44529556/75600707-150f4b80-5af6-11ea-848b-594ecd012307.png)
`cost = torch.mean((hypothesis - y_train)**2)`

**Gradient descent**
Stochastic gradient descent(SGD)
![image](https://user-images.githubusercontent.com/44529556/75600794-2f95f480-5af7-11ea-9d4a-6e6171165e80.png)
![image](https://user-images.githubusercontent.com/44529556/75600824-99160300-5af7-11ea-887d-342fae1f4960.png)


![image](https://user-images.githubusercontent.com/44529556/75600838-bea30c80-5af7-11ea-8fdb-1bcd7a6082f9.png)

```
import torch

x_train = torch.FloatTensor([[1],[2],[3]])
y_train = torch.FloatTensor([[2],[4],[6]])

#requires_grad = True -> 학습할 것이라고 명시
W = torch.zeros(1, requires_grad=True)
b = torch.zeros(1, requires_grad=True)

optimizer = torch.optim.SGD([W,b], lr = 0.01)

nb_epochs = 1000
for epoch in range(1, nb_epochs + 1) :
  hypothesis = x_train * W + b
  cost = torch.mean((hypothesis - y_train)**2)

  optimizer.zero_grad()
  cost.backward()
  optimizer.step()

print(W, b)
```

**Comment by mingxoxo** [comment]
![image](https://user-images.githubusercontent.com/44529556/75600916-da5ae280-5af8-11ea-8aca-25e03b56b0e9.png)
